temporal-sdk/src/main/java/io/temporal/internal/client/ScheduleProtoUtil.java

```diff
@@ -0,0 +1,477 @@
+    if ((scheduleClientContextPropagators.isEmpty() && scheduleOptionsContextPropagators == null)
+        || (scheduleOptionsContextPropagators != null
+            && scheduleOptionsContextPropagators.isEmpty())) {
+      return null;
+    }
+
+    List<ContextPropagator> listToUse =
+        MoreObjects.firstNonNull(
+            scheduleOptionsContextPropagators, scheduleClientContextPropagators);
+    Map<String, Payload> result = new HashMap<>();
+    for (ContextPropagator propagator : listToUse) {
+      result.putAll(propagator.serializeContext(propagator.getCurrentContext()));
+    }
+    return new io.temporal.common.interceptors.Header(result);
+  }
+
+  public ScheduleAction actionToProto(io.temporal.client.schedules.ScheduleAction action) {
+    if (action instanceof ScheduleActionStartWorkflow) {
+      ScheduleActionStartWorkflow startWorkflowAction = (ScheduleActionStartWorkflow) action;
+
+      WorkflowOptions wfOptions = startWorkflowAction.getOptions();
+      // Disallow some options
+      if (wfOptions.getWorkflowIdReusePolicy() != null) {
+        throw new IllegalArgumentException(
+            "ID reuse policy cannot change from default for scheduled workflow");
+      }
+      if (wfOptions.getCronSchedule() != null) {
+        throw new IllegalArgumentException("Cron schedule cannot be set on scheduled workflow");
+      }
+      // Validate required options
+      if (wfOptions.getWorkflowId() == null || wfOptions.getWorkflowId().isEmpty()) {
+        throw new IllegalArgumentException("ID required on workflow action");
+      }
+      if (wfOptions.getTaskQueue() == null || wfOptions.getTaskQueue().isEmpty()) {
+        throw new IllegalArgumentException("Task queue required on workflow action");
+      }
+
+      NewWorkflowExecutionInfo.Builder workflowRequest =
+          NewWorkflowExecutionInfo.newBuilder()
+              .setWorkflowId(wfOptions.getWorkflowId())
+              .setWorkflowType(
+                  WorkflowType.newBuilder().setName(startWorkflowAction.getWorkflowType()).build())
+              .setWorkflowRunTimeout(
+                  ProtobufTimeUtils.toProtoDuration(wfOptions.getWorkflowRunTimeout()))
+              .setWorkflowExecutionTimeout(
+                  ProtobufTimeUtils.toProtoDuration(wfOptions.getWorkflowExecutionTimeout()))
+              .setWorkflowTaskTimeout(
+                  ProtobufTimeUtils.toProtoDuration(wfOptions.getWorkflowTaskTimeout()))
+              .setTaskQueue(TaskQueue.newBuilder().setName(wfOptions.getTaskQueue()).build());
+
+      startWorkflowAction.getArgs().setDataConverter(clientOptions.getDataConverter());
+      Optional<Payloads> inputArgs = startWorkflowAction.getArgs().toPayloads();
+      if (inputArgs.isPresent()) {
+        workflowRequest.setInput(inputArgs.get());
+      }
+
+      if (startWorkflowAction.getOptions().getMemo() != null) {
+        Map<String, Payload> memo = new HashMap<>();
+        for (Map.Entry<String, Object> item :
```

Comment:
Is there not a common util somewhere that does this memo construction? Seems like that might be worth having.

Reply:
This logic  is custom to schedule because the options can get repeatedly decoded/encoded